scrapy_spider_metadata/_params.py

```diff
@@ -0,0 +1,88 @@
+from enum import Enum
+from typing import Any, Dict, Generic, TypeVar
+
+from pydantic import BaseModel
+
+from ._utils import get_generic_param
+
+
+def _unwrap_allof(value, defs, /):
+    allof = value.pop("allOf", None)
+    if allof is None:
+        return
+    for entry in allof:
+        ref = entry.pop("$ref", None)
+        if ref:
+            def_id = ref.rsplit("/", maxsplit=1)[1]
+            entry.update(defs[def_id])
+        entry.pop("title", None)
+        value.update(entry)
+
+
+def _normalize_enum_meta_keys(value, /):
+    enum_meta = value.get("enumMeta", None)
+    if not enum_meta:
+        return
+    for key in list(enum_meta.keys()):
+        if not isinstance(key, Enum):
+            continue
+        enum_meta[key.value] = enum_meta.pop(key)
+
+
+def _update_old_pydantic(value, /):
+    extra = value.pop("json_schema_extra", None)
+    if extra:
+        value.update(extra)
+
+    if "enum" in value:
+        default = value.get("default", None)
+        if default is not None:
+            value["default"] = default.value
+
+
+def _post_process_param_schema(param_schema):
+    defs = param_schema.pop("$defs", None)
+    params = param_schema.get("properties", None)
+    if not params:
+        return
+    for value in params.values():
+        _update_old_pydantic(value)
+        _unwrap_allof(value, defs)
+        _normalize_enum_meta_keys(value)
+
+
+ParamSpecT = TypeVar("ParamSpecT")
+
+
+class Parameterized(Generic[ParamSpecT]):
+    """Validates and type-converts :ref:`spider arguments <spiderargs>` into
+    the :attr:`args` instance attribute according to the :ref:`spider parameter
+    specification <define-params>`.
+    """
+
+    def __init__(self, *args, **kwargs):
+        param_model = get_generic_param(self.__class__, Parameterized)
+        #: :ref:`Spider arguments <spiderargs>` parsed according to the
+        #: :ref:`spider parameter specification <define-params>`.
+        assert param_model is not None
+        self.args: BaseModel = param_model(**kwargs)
```

Comment:
Is it possible to annotate self.args as ParamSpecT, maybe at the top level?

Reply:
I [gave it a first try](https://github.com/scrapy-plugins/scrapy-spider-metadata/pull/1/commits/4d28a3481b9d95e71e9281e3c26359030501ad13), and got:

```
/home/adrian/temporal/scrapy-spider-metadata/tests_typing/test_params.mypy-testing:23: note (mismatch):
  A: note: Revealed type is "Any"
  E: note: Revealed type is 'ParamSpider'
                         ^
/home/adrian/temporal/scrapy-spider-metadata/tests_typing/test_params.mypy-testing:24: note (mismatch):
  A: note: Revealed type is "Any"
  E: note: Revealed type is 'Params'
```

Any idea what could be the issue? I wonder if the lack of `py.typed` in Scrapy could be the cause.

Reply:
Oh we don't have it? We should add it.

Reply:
It turns out the issue was on my end, [`get_spider` needed to be typed](https://github.com/scrapy-plugins/scrapy-spider-metadata/pull/1/commits/18ad10254655b59d7c5325011095f6cadb957fd6).